Use `cc` to build `liblzma`; import `xz` as a sumbodule

abonander opened the pull request:
Converts `lzma-sys/build.rs` to use `cc` to build `xz/src/liblzma` directly instead of xz's autoconf/make build system. Imports `xz` as a submodule directly from the Tukaani git.

closes #29

abonander commented:
While I'm at it, it might be worth while to pare the files from `xz` down to just the bare essentials to make downloads of `lzma-sys` faster. The bloat is definitely non-negligible.

alexcrichton commented:
Thanks! This looks good to me, but CI configs may need an  update to checkout submodules?

abonander commented:
@alexcrichton I reverted the submodule change to keep things simple. I'm having trouble accessing the build log but it should be the same failures as I was seeing locally, every test falling on unwrapping an `Optipms` error.

alexcrichton commented:
Ah currently it looks like:

```
--- stdout
cargo:rerun-if-env-changed=LZMA_API_STATIC
cargo:rustc-link-search=/home/travis/build/alexcrichton/xz2-rs/target/debug/build/lzma-sys-67a7d0e91f4a2092/out/lib
cargo:root=/home/travis/build/alexcrichton/xz2-rs/target/debug/build/lzma-sys-67a7d0e91f4a2092/out
cargo:include=/home/travis/build/alexcrichton/xz2-rs/target/debug/build/lzma-sys-67a7d0e91f4a2092/out/include
cargo:rerun-if-changed=xz/configure
--- stderr
thread 'main' panicked at 'failed to read dir xz/src/liblzma/common: Os { code: 2, kind: NotFound, message: "No such file or directory" }', libcore/result.rs:945:5
```

abonander commented:
Yeah I thought I fixed that path when I reverted the submodule change but it must have gotten lost in the rebase. Now I'm wondering why two of the Travis jobs are green.

abonander commented:
Does Travis not have a libc with Pthreads installed? I'm passing both `-pthread` and `-lpthread` to `cc` and still getting "undefined identifiers" for Pthread types/functions.

alexcrichton commented:
This may be something like `-D_GNU_SOURCE` or something like that to get posix extensions maybe?

abonander commented:
@alexcrichton I don't think that's the issue, this is the function missing: http://man7.org/linux/man-pages/man3/pthread_sigmask.3.html

But `SIG_SETMASK` is undefined as well and that's not specifically pthreads, it's also used with `sigprocmask()`.

It's not an import error and not a pthreads linking error, it includes both `pthread.h` and `signal.h` just a few lines up in that file and uses a few types from pthreads as well. It's specifically this signals API but I don't understand why it's not available. We're using C99 and `pthread_sigmask` has a `_POSIX_C_SOURCE` of `>= 199506L`. 

I'm trying to search for the definitions of these items to see if there's some sort of condition on their exports.

abonander commented:
I see, I do need to manually define `__POSIX_C_SOURCE`.

abonander commented:
That should fix the Linux build, still gotta fix the actual failing tests but that's progress. Gotta fix MSVC14 now.

alexcrichton commented:
It looks like linux compiled (yay!) but I think it failed at runtime?

abonander commented:
Yeah those are the same panics that I was seeing on MSVC16/VS2017 so it's at least consistent with that. I'll have to figure out the cause of course but I imagine it's just missing defaults for some config values or something.

The compile failure on Appveyor is much more interesting. Seems like something with unsupported syntax.

abonander commented:
Nevermind, I somehow managed to fix the builds on both. Now just gotta look at those runtime errors. 

@alexcrichton we probably should add MSVC17 to the AppVeyor build matrix and OS X builds with `LZMA_API_STATIC=1` to Travis-CI.

abonander commented:
That fixes most of the failing tests, just one remaining.

alexcrichton commented:
Nice! 

I'd be totally fine btw adding more entries to the matrix as well

abonander commented:
@alexcrichton that commit should push the build green, though I don't understand the purpose of the SPARC/x86/ARM filters or why XZ only provides tests for SPARC and x86 encode/decode. I've only enabled the bare minimum but there may be other encoders/decoders we want to enable, ARM especially.

abonander commented:
I rewrote the Travis matrix but Travis isn't liking it so I've gotta install the Travis Ruby gem so I can lint it. I see the failure on the 3rd last commit, more missing defines I suspect.

abonander commented:
Travis is running again, but it's not installing `lzma-dev` like it's supposed to.

alexcrichton commented on .travis.yml:
```diff
@@ -1,27 +1,29 @@
+  - nightly
+os:
+  - linux
+  - osx
```
I'd personally prefer to not do this as we don't really need to run stable/beta/nightly on OSX,  could the previous iteration of `include:` be used instead with explicit additions if necessary?

abonander replied:
It seems simpler to just exclude beta and nightly on OS X with this matrix.

alexcrichton replied:
BTW I'd still prefer to not do this, but if you'd like I can take care of switching it back.

abonander replied:
I've got it excluding beta and nightly below this, it's only running two OS X jobs.

alexcrichton replied:
Ah yeah sure, I just prefer a bunch of explicit `include:` lines as historically I've found working with a matrix doesn't really work well in the long run (I don't mind cleaning this up to my own personal linking, which may be a bit odd, though)

abonander replied:
I think this is nicer than manually declaring 8 jobs but it's your repo. I can fix it if you prefer.

abonander commented:
@alexcrichton I'm trying to get `systest` to pass, I'm printing `cargo:include` from the build script so it picks up the include path for `lzma.h` but I'm still getting an error:

```
running: "C:\\Program Files (x86)\\Microsoft Visual Studio\\2017\\Community\\VC\\Tools\\MSVC\\14.13.26128\\bin\\HostX64\\x64\\cl.exe" "/nologo" "/MD" "/Z7" "/I" "\\\\?\\C:\\Users\\cyber\\Rust\\xz2-rs\\lzma-sys\\xz-5.2.3\\src\\liblzma\\api" "/W4" "/W3" "/Wall" "/WX" "/wd4820" "/wd4100" "/wd4996" "/wd4296" "/wd4255" "/wd4668" "/DLZMA_API_STATIC" "/FoC:\\Users\\cyber\\Rust\\xz2-rs\\target\\debug\\build\\systest-40c05e37e47efc8b\\out\\all.o" "/c" "C:\\Users\\cyber\\Rust\\xz2-rs\\target\\debug\\build\\systest-40c05e37e47efc8b\\out\\all.c"
cargo:warning=cl : Command line warning D9025 : overriding '/W4' with '/W3'
all.c
C:\Users\cyber\Rust\xz2-rs\target\debug\build\systest-40c05e37e47efc8b\out\all.c(3): fatal error C1083: Cannot open include file: 'lzma.h': No such file or directory
```

The include is being passed correctly but it's still not picking up the header file. What gives?

abonander commented:
@alexcrichton Running into the issue of `lzma_physmem` and `lzma_cpucores` again. The only place these are used are in `systest` but removing them from `lzma-sys` is a breaking change. You okay with that or should I try to get them working?

abonander commented:
Fixed the include issue. MSVC apparently doesn't like the extended-length path syntax so instead of `Path::canonicalize()` I just joined it with the current dir.

abonander commented:
That should (hopefully) be everything. Once the build is green I'll squash to three commits:

* the changes to `lzma-sys/build.rs`
* the changes to AppVeyor
* the changes to Travis

Additionally I can go back again and look at making `xz` a submodule, it's looking like it's actually not really necessary to patch it because I'm not seeing errors on those `DWORD` casts anymore. It's still warning, but not erroring, and casting pointer types doesn't really affect anything.

abonander commented:
@alexcrichton Looking back at the original `.travis.yml` it seems that the intention was to build and link liblzma whether or not `LZMA_API_STATIC` was defined; there was an extra job in `matrix:include` that installed `lzma-dev` and only ran `cargo test` and `cargo test --features tokio`.

Was this intended? The gist is that `systest` will not run on the job with preinstalled liblzma; it works okay on my WSL but Travis gets some undeclared identifier errors.

alexcrichton commented:
Ah the original intention was to basically just test against linking to the system lzma, ideally `systest` would run but it's not required.

This is looking good though! Did you need some help with the final errors?

abonander commented:
I'll look more closely later this evening, one thing that seems to be going wrong on Travis is that it's running the `pre_script` and `after_success` stages on more than one job which is not what we want. I guess the `stages` conditions can't match on the values of the `rust` key so I'll have to use the environment var that Travis sets instead.

Otherwise, I don't quite understand what's going wrong with `systest` on those jobs since liblzma should be getting built correctly. I imagine it's more missing defines (gotta love stringly typed configuration through the C preprocessor) but it'll take some sifting through liblzma's headers to figure it out. I noticed the `/I` argument isn't being passed to `cc` in this case either.

abonander commented:
@alexcrichton can you rerun the Appveyor jobs? All the failures seem to have been timeouts with crates.io.

alexcrichton commented:
I've hit the rebuild button

abonander commented:
The conditionals are working correctly on Travis now but there's still the same error.

abonander commented:
I feel like this error could be caused by it linking an older version of `lzma-dev` that doesn't have these constants. This is consistent with `cc` not being given an include directory but finding `lzma.h` anyway. However, I thought I was also seeing this error when I had `lzma-dev` installed on all jobs which should have brought it up to date.

abonander commented:
@alexcrichton so at some point someone confused `lzma-dev` and `liblzma-dev` because the latter is what we're actually linking here; installing the former appears to have no function. Presumably `liblzma-dev` comes preinstalled with the standard Ubuntu distros.

Travis-CI uses Ubuntu Trusty which has [an older version of `liblzma` than we're building against](https://packages.ubuntu.com/search?keywords=liblzma-dev); there is no source release (of `5.1.1alpha`) that I can find on Tukaani's site, but 5.0.8 does *not* contain the missing identifiers so we can safely assume they were added in 5.2,

This also explains why I was confused because I have it working locally with the same build steps, but I have `liblzma-dev` already installed on my WSL-running-Bionic: version 5.2.2.

Since we can't add PPAs or anything without `sudo`, I guess for `!LZMA_API_STATIC` we can just skip `systest` which should remedy the issue.

alexcrichton commented:
I think it's fine yeah to avoid running `systest` against system builds, sorry for the headaches you've been getting here!

Additionally I think that liblzma has been ABI-stable for some time, so it shouldn't be too much of a problem with mismatched versions? (in that you'll get link errors or it works)

abonander commented:
My main concerns were:

* variant `LZMA_FULL_BARRIER` added to enum `lzma_action`
* constant `LZMA_IGNORE_CHECK` added to `container.h`

However, it appears that neither is all that important and if either is passed to a `liblzma` that doesn't recognize it, it will just be ignored.

abonander commented:
I amended the commit to fix the formatting in `.travis.yml` real quick but before I pushed it, the first job of the build went green which was failing before, so this should finally be about ready to go! I'll squash when everything is green.

abonander commented on lzma-sys/xz-5.2.3/src/liblzma/rangecoder/range_common.h:
```diff
@@ -14,10 +14,7 @@
-#	include "common.h"
-#endif
-
+#include "common.h"
```
@alexcrichton this change is actually necessary to get `liblzma` to build on its own. Do you want me to commit a patch to make it easy to re-fix on upgrades?

Defining `HAVE_CONFIG_H` is an okay-ish substitute, we'd have to write a `config.h` with the defines it's expecting.

alexcrichton replied:
Hm I wonder, in the interest of eventually using a submodule, could we avoid this? We could write an empty `config.h` to get included, right?

abonander replied:
Yeah, I figured that'd be the preferred approach, but if we're gonna have to write a `config.h`, we might as well stick all the relevant defines in there anyway.

abonander commented:
@alexcrichton I remade `xz` into a submodule and created `config.h` to hold all the defines. It builds locally on Windows and Linux so hopefully the build is green.

abonander commented:
Ah, Appveyor doesn't automatically checkout submodules I see. That's inconvenient, but fixable.

abonander commented:
That oughta do it.

abonander commented on .gitmodules:
```diff
@@ -1,3 +1,7 @@
 [submodule "xz2-sys/xz"]
 	path = xz2-sys/xz
 	url = http://git.tukaani.org/xz.git
```
@alexcrichton Want me to get rid of this submodule entry? It doesn't do anything since  the submodule directory isn't checked-in anymore.

alexcrichton commented:
This looks great to me, thanks so much!
